tests/hikari/test_files.py

```diff
@@ -202,3 +205,83 @@ async def test_test_context_manager_when_target_is_dir(self):
+        return messages.Attachment(
+            id=123,
+            filename="file.txt",
+            media_type="text",
+            height=None,
+            width=None,
+            proxy_url="htt",
+            size=len(b"line1\nline2\nline3\n"),
+            url="https://rick.roll",
+            is_ephemeral=False,
+        )
+
+    async def create_stream(self):
+        # https://github.com/aio-libs/aiohttp/blob/master/tests/test_streams.py#L26
+        loop = asyncio.get_event_loop()
+        protocol = mock.Mock(_reading_paused=False)
+        stream = aiohttp.streams.StreamReader(protocol, 2**16, loop=loop)
+        stream.feed_data(b"line1\nline2\nline3\n")
+        stream.feed_eof()
+        return stream
+
+    async def create_reader(self):
+        return files.WebReader(
+            stream=await self.create_stream(),
+            url="https://rick.roll",
+            status=200,
+            reason="None",
+            filename="file.txt",
+            charset=None,
+            mimetype="your mum",
+            size=len(b"line1\nline2\nline3\n"),
+            head_only=False,
+        )
+
+    @pytest.mark.asyncio()
+    async def test_resource_read(self, attachment: messages.Attachment, data: bytes):
+        with mock.patch.object(files._WebReaderAsyncReaderContextManagerImpl, "__aenter__") as mock_aenter:
+            mock_aenter.return_value = await self.create_reader()
+
+            # Mocking the previous object throws an error here
+            # so we just want to avoid it.
+            try:
+                assert await attachment.read() == data
+            except AttributeError as exc:
+                if str(exc) != "'NotImplementedType' object has no attribute '__aexit__'":
+                    raise
+
+    @pytest.mark.asyncio()
+    async def test_resource_save(self, attachment: messages.Attachment, data: bytes):
+        with mock.patch.object(files._WebReaderAsyncReaderContextManagerImpl, "__aenter__") as mock_aenter:
+            mock_aenter.return_value = await self.create_reader()
+
+            # A try, finally is used to delete the file rather than relying on delete=True behaviour
+            # as on Windows the file cannot be accessed by other processes if delete is True.
+            try:
+                file = tempfile.NamedTemporaryFile("wb", delete=False)
+                path = pathlib.Path(file.name)
+
+                with file:
+                    # Mocking the previous object throws an error here
+                    # so we just want to avoid it.
+                    try:
+                        await attachment.save(path=path)
+                    except AttributeError as exc:
+                        if str(exc) != "'NotImplementedType' object has no attribute '__aexit__'":
+                            raise
+
+                    with open(path, "rb") as f:
+                        assert f.read() == data
+            finally:
+                path.unlink()
```

Comment:
You can make use of mocking here instead of writing to disk and having actual objects. Have a look at #1273 for example of this.

Furthermore, instead of creating an Attachment object, have a fixture like so:
```py
@pytest.fixture
def resource():
    class ResourceImpl(files.Resource):
        stream = mock.Mock()
        url = "https://some.where/file.txt"
        filename = "file.txt"

     return ResourceImpl()
```

I was going to write an example test for this to help you get started, but haven't been able to finish it. Lmk if you need any help

Reply:
I'm not really sure tests you want me to do instead of these. I've looked at #1273 but am not getting a clear idea of how you want me to adapt them for this PR.

Reply:
There are 2 things:

1. Instead of creating an attachment object, you can just create a resource impl to use (codeblock above).
2. Instead of writing to disk, you can mock and make sure the correct things are called: https://github.com/hikari-py/hikari/blob/612b610fcf06bb1b335c02a7a098b2a1c7c1a0b2/tests/hikari/test_files.py#L74-L95

Reply:
I understand that, I just can't visualise in my head *how* those would be achieved. Like I don't know how you'd probably test the features using mocks instead of writing to the disk.